Implemented #83 (HTTPHeaders accessors)

amosavian opened the pull request:
This pull request includes static variables for `HTTPVersion` struct. Also implemented easy accessors for `HTTPHeaders` to strict types for header values.
~`Authentication`, `Cookie` and `Set-Cookie` accessors are not implemented yet due to complexity or designing limitations. I will make another PR when they became ready in near future.~

**UPDATE:** All methods are implemented now. Here is mailing list post: mailing list: https://lists.swift.org/pipermail/swift-server-dev/Week-of-Mon-20171113/000713.html

amosavian committed:
Implementing HTTPVersion static variables

amosavian committed:
Partialy implemented #83 (easy accessors for HTTPHeaders)

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+extension HTTPHeaders {
+    
+    /// MIME type directive for `Accept` and `Content-Type` headers
+    public struct ContentType: RawRepresentable, Hashable, Equatable {
```
This is weird because it can't carry arguments. Maybe this should be called MediaType? And then a ContentType for a content-type?

E.g. how do you do:

    Content-Type: multipart/form-data; boundary=something

or

    Content-Type: text/calendar; component=VCARD

amosavian replied:
I implemented `contentCharset` property for `charset` parameter. We can change design btw

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+    }
+    
+    /// Encoding of body
+    public enum Encoding: String {
```
This needs to be open, an `enum` is not appropriate here for obvious reasons. (at the minimum add a `.other`, but that is not really neat either - rather do constants)

amosavian replied:
Encoding should be supported by server. I believe we can't have custom values when they are not supported unless you show me I'm wrong.

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+    // Request Headers
+    
+    /// Fetch `Accept` header values, sorted by `q` parameter
+    public var accept: [ContentType] {
```
What about wildcard support? Quality arguments? Supporting this would be the requirement for a structured header:

    Accept: text/html, application/xhtml+xml, application/xml;q=0.9, */*;q=0.8

amosavian replied:
- This property is a getter. setter method(`set(accept:, quality:)`) take quality as a argument, and getter method sorts values. I believe a server implementation doesn't need q value when It's sorted already.
- There is a `ContentType.all` which is wildcard, there are also wildcards for audio, video and text.

helje5 replied:
This is intended to be not just for the server, just fwiw.

amosavian replied:
Yes, I implemented getter and setter method for all headers. They're maybe redundant in some cases.

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+    // TODO: Implement var cookie: [HTTPCookie]
+    
+    /// `If-Match` header etag value
+    public var ifMatch: String? {
```
If-Match takes an array of etags, supports weak and wildcards. For a typed header this seems very insufficient.

amosavian replied:
It returns eTags with quotations. so you can have `"W/\"12345\""` as an etag. Maybe we can present value as an associated enum? (e.g. `.strong("12345")`, `.weak("12345")` and `.wildcard("*")`)
We can change design.

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+    }
+    
+    /// `Content-Length` header value
+    public var contentLength: Int64? {
```
I'm not convinced `Int64` is right here as you can't expect a x32 implementation to deal w/ x64 Integers. I think that should be rejected at the HTTP level with a proper response.

amosavian replied:
Apple's `URLResponse.expectedContentLength` is `Int64`, I decided to use Int64 for compatibility. A file can be bigger than 2TB btw.

helje5 commented on Sources/HTTP/HTTPVersion.swift:
```diff
@@ -20,6 +20,17 @@ public struct HTTPVersion {
 
+extension HTTPVersion {
+    /// HTTP version 1.0
+    public static let v1_0 = HTTPVersion(major: 1, minor: 0)
```
I don't know, maybe, why not. I'm not sure when I would ever use those constants.

dschwartz783 replied:
Maybe someone has an old copy of Netscape lying around, and can't use HTTP 1.1. :(

amosavian replied:
HTTP 1.0 and 1.1 are different only in default values for `Keep-Alive` and `Connection` headers afaik. We will replace instances in current code like [here](https://github.com/swift-server/http/blob/31d066a0e92b4575a4389d592e022fe0b9933956/Sources/HTTP/HTTPStreamingParser.swift#L401)

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+        let fm = DateFormatter()
+        fm.dateFormat = standard.rawValue
+        fm.timeZone = timeZone ?? TimeZone(identifier: "UTC")
+        fm.locale = locale ?? Locale(identifier: "en_US_POSIX")
```
The timezone and locale should at least get cached.

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+    /// Checks date string against various RFC standards and returns `Date`.
+    init?(rfcString: String) {
+        let dateFor: DateFormatter = DateFormatter()
+        dateFor.locale = Locale(identifier: "en_US")
```
cache

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+    }
+    
+    /// `ETag` header value
+    public var eTag: String? {
```
the etag header contains an array. also, no support for w/

amosavian replied:
ditto from `If-Match`

helje5 commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,795 @@
+        set {
+            if let newValue = newValue {
+                // eTag value must be a quoted string, we check only suffix to be compatible with weak eTags
+                let finalValue = newValue.hasSuffix("\"") ? newValue : "\"\(newValue)\""
```
escaping?

amosavian replied:
No, that adds a quotation to newValue if it doesn't have already. for weak etags

helje5 reviewed (commented):
While I'm kinda think that we should have typed headers, I think this PR has tons of issues and doesn't reflect the RFCs.

amosavian committed:
Strict-typed EntryTag

amosavian committed:
Cached locale/timezone for Date, contentRange setters enhanced

helje5 commented:
May I also suggest that you take this to the mailing list for discussion. I think we already discussed this kind of stuff and decided not to do typed headers at this level of the framework. It would be part of a real framework on top of this driver.

If we really go the route of typed headers (which I personally kinda like for efficiency), we should actually do that in the storage itself, not just wrapping the String key/value. I wrote some email on that in the list.

amosavian commented:
@helje5 please check new implementation of EntryTag.

amosavian commented:
I don't know how to use mailing list unfortunately. I wonder if you could help me.

Headers are not something that you would access frequently. For requests, you may access them once or twice. For response, you set them once. Parsing all request headers when initializing is not an optimized behavior unless they got accessed (lazy). We may introduce some kind of caching later btw.

helje5 commented:
https://lists.swift.org/mailman/listinfo/swift-server-dev

amosavian committed:
Fixed converting IANA to String.Encoding on linux

amosavian committed:
Patch: Fixed converting IANA to String.Encoding on linux

amosavian commented:
@helje5 I sent email and it's awaiting moderator approval

helje5 commented:
You first need to subscribe to the list with the email address you want to use for sending messages to it. Once this is through, you don't need to wait for moderator approval.

amosavian committed:
added contentDisposition, refactored CharacterSet.quoted

amosavian committed:
Default charset is isoLatin1 according to HTTP, code linted

amosavian commented:
Posted in mailing list: https://lists.swift.org/pipermail/swift-server-dev/Week-of-Mon-20171113/000713.html

amosavian committed:
Content-Disposition filename* charset consideration, refactorings

amosavian committed:
Fixed HTTPHeaders.Encoding, lint

amosavian committed:
Added Authorization, Cookie, TE, Connection, Trailer accessors

amosavian committed:
Added total size parameter to set(contentRange:)

amosavian committed:
Realm param is quoted, Added documentations

nacho4d commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,1229 @@
+    /// Determines server accepts `Range` header or not
+    public enum RangeType: String {
+        /// Can't accept Range
+        case none
```
I recently read RFC7233 for implementation somewhere else. RCF allows to define any kind of range not only `bytes`. In the specific cases of File Serving `bytes` just happens to be the most common case. Ranges could be used in other REST, etc calls too and it could be `items`, `rows`, etc whatever, restricting this to only bytes might not allow http module users to use this function to other purposes than file serving.

amosavian replied:
Please find fixed RangeType and ifRange implementations

nacho4d commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,1229 @@
+    /// `If-Range` header etag value
+    public var ifRange: EntryTag? {
+        get {
+            return self.storage[.ifRange]?.first.flatMap(EntryTag.init)
```
As of RFC7233 and RFC7232 `ifRange` could also be of the form of `Last-Modified` header. instantiating from EntryTag will not allow the later form.

amosavian committed:
RangeType became struct, fixed some bugs

amosavian committed:
Added MediaType & params to ContentType, Fixed If-Range

amosavian committed:
Added Range & SetCookie get accessors, Added clientBrowser & clientOperatingSystem

amosavian committed:
Implemented filename* for utf8 Content-Disposition file names

amosavian committed:
Fixed linux substring compilation error for trimmingCharacters()

amosavian committed:
Implemented add(setCookie:) as Set-Cookie setter, few optimizations

amosavian committed:
Added documentations, cacheControl returns array and supports custom values
- Better browser detection
- Added `ContentType.xhtml` and removed unusable ones.

nacho4d commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,1596 @@
+    
+    /// Returns `Range` header value.
+    /// - Note: upperbound will be Int64.max in case of open ended Range.
+    public var range: Range<Int64>? {
```
According to RFC7233 range headers can have multiple ranges like: `Range: bytes=3-5,10-15`. So this should be defined as `[Range<Int64>]?`.

Also, the spec for parsing ranges is not as simple as below 4 lines but I guess that can be addressed later as long as the interface is ok :p

amosavian replied:
As far as I understood from examples in RFC, multiple ranges are continuous if we intersect values. e.g:

`bytes=0-0,-1` becomes: -1..<1
`bytes=500-600,601-999` becomes: 500..<1000
`bytes=500-700,601-999` becomes: 500..<1000

Thus it is not an array of range. Please correct me if I'm wrong?

Update: RFC explicitly says ` byte-range-set  = 1#( byte-range-spec / suffix-byte-range-spec )` which means one value is acceptable, and calls multiple ranges as "valid (but not canonical)". All examples are continuous though.

nacho4d replied:
Spec defines various ranges, they do not have to be continuous, for example response in page 11 shows at least two ranges bytes=500-999,7000-7999

As per spec, server CAN merge overlapping or adjacent ranges however that is not a must. Multipart response for partial content is clearly defined which indicates the existence of multiple ranges. 

     HTTP/1.1 206 Partial Content
     Date: Wed, 15 Nov 1995 06:25:24 GMT
     Last-Modified: Wed, 15 Nov 1995 04:58:08 GMT
     Content-Length: 1741
     Content-Type: multipart/byteranges; boundary=THIS_STRING_SEPARATES

     --THIS_STRING_SEPARATES
     Content-Type: application/pdf
     Content-Range: bytes 500-999/8000

     ...the first range...
     --THIS_STRING_SEPARATES
     Content-Type: application/pdf
     Content-Range: bytes 7000-7999/8000

     ...the second range
     --THIS_STRING_SEPARATES--


https://tools.ietf.org/html/rfc7233

nacho4d replied:
Spec defines various ranges, they do not have to be continuous, for example response in page 11 shows at least two ranges bytes=500-999,7000-7999

As per spec, server CAN merge overlapping or adjacent ranges however that is not a must. Multipart response for partial content is clearly defined which indicates the existence of multiple ranges. 

     HTTP/1.1 206 Partial Content
     Date: Wed, 15 Nov 1995 06:25:24 GMT
     Last-Modified: Wed, 15 Nov 1995 04:58:08 GMT
     Content-Length: 1741
     Content-Type: multipart/byteranges; boundary=THIS_STRING_SEPARATES

     --THIS_STRING_SEPARATES
     Content-Type: application/pdf
     Content-Range: bytes 500-999/8000

     ...the first range...
     --THIS_STRING_SEPARATES
     Content-Type: application/pdf
     Content-Range: bytes 7000-7999/8000

     ...the second range
     --THIS_STRING_SEPARATES--


https://tools.ietf.org/html/rfc7233

amosavian committed:
Fixed `range` to return array, Added `acceptLanguage`

amosavian committed:
HTTPHeaders as struct

nacho4d commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,1677 @@
+    /// - Note: upperbound will be Int64.max in case of open ended Range.
+    /// - Note: Server response should be `multipart/byteranges` in case of more than one range is returned.
+    ///    See [MDN's HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) for more info.
+    /// - Important: A negative range means server must return last bytes/items of file.
```
`bytes` is the most common type and to validate ranges one would get the file size from a file as: `(fileAttributes[FileAttributeKey.size] as? NSNumber)?.uint64Value` because file size are declared as `UInt64` in Foundation. However current declaration is `Int64` which means there would be cases when the string can not be converted to Range<Int64>. How can we address this?. 

Example: try to get a tiny part of a very big file: 
`Range: bytes=18446744073709551600-18446744073709551601`

Are you using `Int64` is because of ranges like: `Range: bytes=-3` which look like negative?. Actually there is no such a thing as negative ranges. In fact they should be converted into a  range object/struct **after** the file size is known, for example if fileSize is `1000` then it would mean: `Range: bytes=997-999` which uses only positive numbers.

I like the idea of having everything statically defined as structs/classes/enums rather than raw strings, but I am not sure having this interface will help to correctly implement the RFC :(

(As a side note, nodejs implementation converts range header string into range objects in a function that receives the fileSize as a parameter.)

amosavian replied:
About Foundation: `HTTPURLResponse.expectedContentLength` and `URLSession` methods return `Int64`.

On negative values, We don't define how to interpret data as we don't know about resource. It's upon higher level framework to convert negative values to meaningful ones or return 416.

About BigInt, Using Decimal will hit performance. We are implementing these method for common cases. A file with that size is not common and higher level framework would have to implement theirs. Should we add limitations to documentation?

nacho4d replied:
OK, if other values are Int64 then I think this should be ok :]
IMO I don't think documentation needs to know about this. Other framework implementers will likely read comments in code

amosavian committed:
Performance optimizations, Accessors to parse coma-separated

amosavian committed:
Added acceptCharset, Refactoring parsing quilified

amosavian committed:
Added host, removed redundant Accept-* & If-* setters

amosavian commented:
@seabaylea @carlbrown Any chance to get merged? I must begin working on tests.

amosavian committed:
RFC5987 for Authorization & WWW-Authenticate
- Refactored ContentType for better performance

amosavian committed:
More flexible Content-Deposition

amosavian committed:
Better RFC 5987 support

aleksey-mashanov commented:
Great work!
Typed headers are definitely the thing a good framework should have. But I have a sense that this too much for an API between an HTTP framework and an HTTP server. Maybe this should be made as a separate module?

I suggest to replace `Int64` with `Int` to follow Swift guidelines:
> Unless you need to work with a specific size of integer, always use `Int` for integer values in your code.

amosavian commented:
@aleksey-mashanov Please find my response for `Int64` [here](https://github.com/swift-server/http/pull/104#discussion_r152649403).  HTTPURLResponse.expectedContentLength and URLSession methods return `Int64`. NSFileManager returns `UInt64` for file size but `Content-Length` can have negative values. Furthermore it's common to have a file larger than 2GB on a 32-bit system.

Please find my explanation [here](https://gist.github.com/amosavian/084636d1a382717b7d7ce7469aaacc07) on why we should implement typed header in this layer. RFC 5987 is a reason e.g.

aleksey-mashanov commented:
>  Furthermore it's common to have a file larger than 2GB on a 32-bit system.

Agree. But 32bit server? We still have some in production for historical reasons but it is a pain. In almost every place in a surrounding code this value will be converted to `size_t` (`Int`) and used for generation of dynamic content. I really don't know what is better: precise API or comfortable API. For low-level of this HTTP API maybe first but this PR is out of scope of low-levelness.

> RFC 5987 is a reason e.g.

Agree. For me RFC5987 and parsing of additional parameters (after `;`) are the most interesting parts of this PR. Maybe they should be implemented lower in `HTTPHeaders` as an additional subscripts?
There are a lot of RFCs, a lot of additional headers and even more will arise. It will be hard to follow all of them correctly and most of changes will be source compatibility breaking.

Again. This PR doesn't provide features which are required to efficiently pass data from a framework to a server and back. Nevertheless it is very useful.

amosavian commented:
> every place in a surrounding code this value will be converted to size_t (Int) 

Swift Foundation's `FileManager` class works with file size as `UInt64`. In case we don't want typecast, it's better to consider `UInt64` rather than `Int`. On the other hand, Swift compiler may give a warning if user tries to cast `UInt64` to `Int64` or `Int`.

Furthermore, `range`, `contentRange` and `contentLength` must have same type imo. `range` must support negative values.

> There are a lot of RFCs, a lot of additional headers and even more will arise.

I mentioned RFC5987 just as an example because it's mentioned in source code as TODO. Definitely There are more cases which I'm not aware of.

>  PR doesn't provide features which are required to efficiently pass data from a framework to a server and back. Nevertheless it is very useful.

This design is implemented as an extension to current design. My primary goal was not to touch current architecture because any change in architecture usually needs extensive workgroup paperworks (PS: I'm not aware how swift-server workgroup works). But my design, which uses initializers for casting instead of functions (except parsing parameters and `q` values), gives the flexibility to change underlying design anytime.

amosavian committed:
Fixed createParam() quotation enclosing bug, Added cookieDictionary
- Small refactors and improvements
- Added more documentation

amosavian committed:
Added Transfer-Encoding accessor

amosavian committed:
Refactored into 2 files

amosavian committed:
Refactored into 2 files, link accessor added

amosavian committed:
Fixed RFC5987 CharacterSet, Fixed accept* accessors
- More refactors

amosavian committed:
Added KeepAlive, Typed relationType for Link, bug fixes
- Added Challenge convenience init
- Fixed parsing range and contentRange
- All types are RawRepresentable now

amosavian committed:
Fix linux build error

amosavian committed:
Added Test cases for HTTPHeaders accessors

amosavian committed:
Improved tests, fixed minor bugs

amosavian committed:
Fix tests list (for HTTPHeaders accessors), small optimizations

amosavian committed:
Fix linux compile error

amosavian committed:
Supporting multiple WWW-Authenticate challenges

amosavian commented:
List of optional but unimplemented headers, which are nice to have in portfolio. I won't implement them as a part of this PR:
`Access-Control-*`, `Accept-Patch`, `Alt-Svc`, `Content-Security-Policy`, `DNT`, `Forwarded`, `Proxy-Authenticate`, `Proxy-Authorization`, `Retry-After`, `Tk`, `Warning`, `X-XSS-Protection`

amosavian committed:
set(contentRange:) supports RangeExpression, Few more tests

amosavian commented:
This PR is completely implemented.

tanner0101 reviewed (approved):
awesome

amosavian commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,800 @@
+        return pairs
+    }
+    
+    /// `Origin` header value.
```
Change to `host`

amosavian replied:
Change comment to `host`

amosavian commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,800 @@
+    }
+    
+    fileprivate func dissectContentRange(_ value: String?) -> (from: UInt64, to: UInt64?, total: UInt64?)? {
+        // Converting real negatives to _ to avoid conflict with - as separator
```
Remove this comment

amosavian commented on Sources/HTTP/HTTPHeadersTypes.swift:
```diff
@@ -0,0 +1,1529 @@
+            switch self {
+            case .basic(let user, let password):
+                let text = "\(user):\(password)"
+                let b64 = (text.data(using: .ascii) ?? text.data(using: .utf8))?.base64EncodedString() ?? ""
```
utf8 is default encoding, change the order

amosavian commented on Sources/HTTP/HTTPHeadersTypes.swift:
```diff
@@ -0,0 +1,1529 @@
+            }
+        }
+        
+        /// Inits a noew
```
Fix comment

amosavian commented on Sources/HTTP/HTTPHeadersTypes.swift:
```diff
@@ -0,0 +1,1529 @@
+        // Images
+        
+        /// All Image types
+        static public let image = MediaType(generalType: "image", type: "bmp")
```
change type: to `”*”`

amosavian commented on Sources/HTTP/HTTPHeadersTypes.swift:
```diff
@@ -0,0 +1,1529 @@
+            switch self {
+            case .strong(let etag):
+                // TODO: Remove non ascii characters
+                let lintedEtag = etag.trimmingCharacters(in: .quotedWhitespace)
```
remove redundant trimming

amosavian commented on Sources/HTTP/HTTPHeadersTypes.swift:
```diff
@@ -0,0 +1,1529 @@
+                let lintedEtag = etag.trimmingCharacters(in: .quotedWhitespace)
+                return "\"\(lintedEtag)\""
+            case .weak(let etag):
+                let lintedEtag = etag.replacingOccurrences(of: "W/\"", with: "", options: [.anchored, .caseInsensitive]).trimmingCharacters(in: .quotedWhitespace)
```
ditto

amosavian commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,800 @@
+        set {
+            if !newValue.isEmpty {
+                // TOCHECK: Forbidden headers ought to be ignored/dropped
+                self.storage[.trailer] = newValue.map { $0.original }
```
replace with `.method`

amosavian commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,800 @@
+        set {
+            if !newValue.isEmpty {
+                // TOCHECK: Forbidden headers ought to be ignored/dropped
+                self.storage[.vary] = newValue.map { $0.original }
```
ditto

amosavian commented on Sources/HTTP/HTTPHeadersTypes.swift:
```diff
@@ -0,0 +1,1529 @@
+        static public let png = MediaType(generalType: "image", type: "png")
+        /// Scalable vector graphics
+        static public let svg = MediaType(generalType: "image", type: "svg+xml")
+        /// Scalable vector graphics
```
Fix comment

amosavian commented on Sources/HTTP/HTTPHeadersAccessors.swift:
```diff
@@ -0,0 +1,800 @@
+    }
+    
+    /// Fetch `Accept-Charset` header values, sorted by `q` parameter. An empty array means no value is set in header.
+    public var acceptCharset: [String.Encoding] {
```
Handling wildcard

amosavian committed:
allowCharset wildcard, added comments
- Better IANA to String.Encoding conversion in Linux

amosavian committed:
Fix linux compile error on ianatable

amosavian committed:
Add more Documentation, Gardening, Renaming

amosavian committed:
Enabling HTTPHeadersAccessors on Linux

amosavian committed:
Fix bugs and testing issues in Linux

amosavian committed:
Generic set(contentRange:), fixed possible crash

amosavian committed:
Fix unsigned int error

mohsen1 commented:
FYI DOM [`Headers`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers) API allows setting multiple values for the same cookie and ignores the key casing:

```js
var h = new Headers()
h.append('Cookie', 'foo bar')
h.append('cooKie', 'baz')
for (var e of h.entries()) { console.log(e) } 
```

prints:

```
["cookie", "foo bar, baz"]
```

ianpartridge commented:
Closing as this project is historical, per the README.
